STCOR-989: Generalise useChunkedCQLFetch hook to allow for other chunked id transformations

EthanFreestone opened the pull request:
## feat: useChunkedIdTransformFetch hook

### Current state
useChunkedCQLFetch is a pre-existing hook which aims to allow for frontend joins, taking in some list of UUIDs and performing a fetch for those objects in a chunked and reliable manner.

### Premise
This hook, as its name suggests, is currently only applicable to endpoints making use of CQL, where the bulk of the internal logic for actually chunking said requests is API agnostic.

Hence this PR aims to generalise that logic, leaving the API for useChunkedCQLFetch in place to avoid backwards compatibility issues, but allowing for frontends pointing at services using other API shapes to make use of the same underlying chunked-fetch logic.


### Use case
In particular, ui-agreements has an upcoming use case where a list of IDs is presented from an external service, and we have to check whether those match anything in our internal DB. The API for this already exists, but the list of IDs can become very long, and the query requires filtering on a couple of fields, so the query string grows very quickly, and regularly returns `414 URI Too Long`.

In particular  the query string is something like:
```?filters=((pci.identifiers.identifier.status == approved && pci.identifiers.identifier.namespace == 'gokb_uuid' && pci.identifiers.identifier == ID_1) || (... ID_2) || (... ID_3))```

### Approach
This PR moves the bulk of the logic which previously lived inside useChunkedCQLFetch up to a new generic hook: `useChunkedIdTransformFetch`, and keeps all CQL-specific logic confined to the old hook (such as the presence of the `query` and `limit` parameters, as well as the specific way to map a list of Ids into such a query)


### Considerations
There is absolutely a case here to suggest that part of the problem is that there ought to be a backend API to handle the namespacing and approved status matching parts, but fundamentally we have an unknown size list of ids coming in, and unless we include those in a HTTP body, the risk of seeing a 414 is there, and the argument about whether it's acceptable to fetch records via some kind of `POST` with a body is a different kettle of fish.

zburke reviewed (commented):
@EthanFreestone, I'm having a hard time gathering what specific changes this PR contains. Neither the PR description nor the Jira ticket say much beyond "generalise the existing component". Can you provide a specific example of how the code is changing, i.e. what kind of "chunked id transformations" are possible now that were not possible before?

EthanFreestone commented:
> @EthanFreestone, I'm having a hard time gathering what specific changes this PR contains. Neither the PR description nor the Jira ticket say much beyond "generalise the existing component". Can you provide a specific example of how the code is changing, i.e. what kind of "chunked id transformations" are possible now that were not possible before?

Heya Zak :)


useChunkedCQLFetch works in exactly the same way as it ever did, however I have a scenario in ui-agreements which requires us to take a list of UUIDs from an external service, and check whether those match anything in our internal DB.


The API for this already exists, but the list of IDs can become quite long, and we need to filter on a couple of fields, so thw query string grows very quickly, easily enough to regularly hit 414s.

The API call chunking logic written for useChunkedCQL is API agnostic, but the query section assumes that all you want to do is fetch all records where the id itself is == a OR b OR c..., and also assumes that the mechanism to filter OR is the same for all APIs, which it is not for KIWT endpoints (agreements/licenses)



In our case I need to append something like `?filters=((pci.identifiers.identifier.status == approved && pci.identifiers.identifier.namespace == 'gokb_uuid' && pci.identifiers.identifier == ID_1) || (... ID_2) || (... ID_3))


This grows very very quickly, and the only thing preventing useChunkedCQLFetch from chunking these calls is the fact that the `query` is assumed and baked in, along with other CQL-only params such as limit.


Instead I have moved the bulk of the logic up and implemented a callback to take a chunked set of ids and perform "some transform", in my case to allow the building of the above filter set per id.


There is _absolutely_ a case here to suggest that part of the problem is that there ought to be a backend API to handle the namespacing and approved status matching parts, but fundamentally we have an unknown size list of ids coming in, and unless we include those in a HTTP body, the risk of seeing a 414 is there.

This also of course allows any other API shape, for any other backends not using CQL

zburke commented:
Noted, @EthanFreestone. I am not deliberately ignoring this. Some time-sensitive security work has sucked up all my free time the last week or two. Hopefully that'll subside now and I'll get to this soon.

EthanFreestone commented on src/queries/useChunkedIdTransformFetch.js:
```diff
@@ -0,0 +1,133 @@
+      queryArray.push({
+        queryKey,
+        // !WARNING! DO NOT PUT LOGIC INTO THE GET HERE
+        // LOGIC FOR CQL FETCHES SHOULD BE IN CHUNKED CQL FETCH HOOK INSTEAD
```
Have changed :)

zburke reviewed (commented):
We're almost there! Just a few more nits to resolve.

EthanFreestone commented on src/queries/useChunkedCQLFetch.js:
```diff
@@ -1,146 +1,18 @@
-  };
+  return useChunkedIdTransformFetch({
+    // This is the place to put logic which performs transforms in CQL queries
+    chunkedQueryIdTransform: (chunkedIds) => `?limit=${limit}&query=${idName}==(${chunkedIds.join(' or ')})`,
```
In this case it's still using the string return, as that's totally valid for UrlSearchParams in JS
